File: NET_8/FoodCall.Domain/Entities/Order.cs
using FoodCall.Domain.Enums;
using FoodCall.Domain.Exceptions;
using FoodCall.Domain.ValueObjects;

namespace FoodCall.Domain.Entities;

public class Order
{
    public Guid Id { get; private set; }
    public Guid CustomerId { get; private set; }
    public Guid RestaurantId { get; private set; }
    public decimal TotalValue { get; private set; }
    public OrderStatus Status { get; private set; }
    public List<OrderItem> Items { get; private set; } = new();

    public Order(Guid customerId, Guid restaurantId)
    {
        ValidateCustomerId(customerId);
        ValidateRestaurantId(restaurantId);

        Id = Guid.NewGuid();
        CustomerId = customerId;
        RestaurantId = restaurantId;
        Status = OrderStatus.Pending;
        TotalValue = 0;
    }

    public void AddItem(Guid productId, string productName, decimal unitPrice, int quantity)
    {
        if (Status != OrderStatus.Pending)
            throw new BusinessRuleException("AddItem", "Só é possível adicionar itens em pedidos pendentes");

        var item = new OrderItem(productId, productName, unitPrice, quantity);
        Items.Add(item);
        RecalculateTotal();
    }

    public void RemoveItem(Guid itemId)
    {
        if (Status != OrderStatus.Pending)
            throw new BusinessRuleException("RemoveItem", "Só é possível remover itens de pedidos pendentes");

        var item = Items.FirstOrDefault(i => i.Id == itemId);
        
        if (item == null)
            throw new EntityNotFoundException("OrderItem", itemId);

        Items.Remove(item);
        RecalculateTotal();
    }

    public void Confirm()
    {
        if (Status != OrderStatus.Pending)
            throw new BusinessRuleException("ConfirmOrder", "Apenas pedidos pendentes podem ser confirmados");

        if (!Items.Any())
            throw new BusinessRuleException("ConfirmOrder", "Pedido deve ter pelo menos um item");

        Status = OrderStatus.Confirmed;
    }

    public void StartPreparing()
    {
        if (Status != OrderStatus.Confirmed)
            throw new BusinessRuleException("StartPreparing", "Apenas pedidos confirmados podem iniciar preparo");

        Status = OrderStatus.Preparing;
    }

    public void SendForDelivery()
    {
        if (Status != OrderStatus.Preparing)
            throw new BusinessRuleException("SendForDelivery", "Apenas pedidos em preparo podem ser enviados");

        Status = OrderStatus.OutForDelivery;
    }

    public void MarkAsDelivered()
    {
        if (Status != OrderStatus.OutForDelivery)
            throw new BusinessRuleException("MarkAsDelivered", "Apenas pedidos em entrega podem ser marcados como entregues");

        Status = OrderStatus.Delivered;
    }

    public void Cancel()
    {
        if (Status == OrderStatus.Delivered)
            throw new BusinessRuleException("CancelOrder", "Pedidos já entregues não podem ser cancelados");

        if (Status == OrderStatus.Cancelled)
            throw new BusinessRuleException("CancelOrder", "Pedido já está cancelado");

        Status = OrderStatus.Cancelled;
    }

    private void RecalculateTotal()
    {
        TotalValue = Items.Sum(i => i.SubTotal);
    }

    private void ValidateCustomerId(Guid customerId)
    {
        if (customerId == Guid.Empty)
            throw new InvalidEntityException("Order", "CustomerId não pode ser vazio");
    }

    private void ValidateRestaurantId(Guid restaurantId)
    {
        if (restaurantId == Guid.Empty)
            throw new InvalidEntityException("Order", "RestaurantId não pode ser vazio");
    }
}

File: Java/src/main/java/com/foodcall/domain/entities/Order.java
package com.foodcall.domain.entities;

import com.foodcall.domain.enums.OrderStatus;
import com.foodcall.domain.exceptions.BusinessRuleException;
import com.foodcall.domain.exceptions.EntityNotFoundException;
import com.foodcall.domain.exceptions.InvalidEntityException;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "orders")
@Getter
@NoArgsConstructor
public class Order {
    
    @Id
    private UUID id;
    
    @Column(nullable = false)
    private UUID customerId;
    
    @Column(nullable = false)
    private UUID restaurantId;
    
    @Column(nullable = false)
    private double totalValue;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatus status;
    
    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "order_id")
    private List<OrderItem> items = new ArrayList<>();
    
    public Order(UUID customerId, UUID restaurantId) {
        validateCustomerId(customerId);
        validateRestaurantId(restaurantId);
        
        this.id = UUID.randomUUID();
        this.customerId = customerId;
        this.restaurantId = restaurantId;
        this.status = OrderStatus.PENDING;
        this.totalValue = 0;
    }
    
    public void addItem(UUID productId, String productName, double unitPrice, int quantity) {
        if (status != OrderStatus.PENDING) {
            throw new BusinessRuleException("AddItem", "Só é possível adicionar itens em pedidos pendentes");
        }
        
        OrderItem item = new OrderItem(productId, productName, unitPrice, quantity);
        this.items.add(item);
        recalculateTotal();
    }
    
    public void removeItem(UUID itemId) {
        if (status != OrderStatus.PENDING) {
            throw new BusinessRuleException("RemoveItem", "Só é possível remover itens de pedidos pendentes");
        }
        
        OrderItem item = items.stream()
                .filter(i -> i.getId().equals(itemId))
                .findFirst()
                .orElseThrow(() -> new EntityNotFoundException("OrderItem", itemId));
        
        this.items.remove(item);
        recalculateTotal();
    }
    
    public void confirm() {
        if (status != OrderStatus.PENDING) {
            throw new BusinessRuleException("ConfirmOrder", "Apenas pedidos pendentes podem ser confirmados");
        }
        if (items.isEmpty()) {
            throw new BusinessRuleException("ConfirmOrder", "Pedido deve ter pelo menos um item");
        }
        this.status = OrderStatus.CONFIRMED;
    }
    
    public void startPreparing() {
        if (status != OrderStatus.CONFIRMED) {
            throw new BusinessRuleException("StartPreparing", "Apenas pedidos confirmados podem iniciar preparo");
        }
        this.status = OrderStatus.PREPARING;
    }
    
    public void sendForDelivery() {
        if (status != OrderStatus.PREPARING) {
            throw new BusinessRuleException("SendForDelivery", "Apenas pedidos em preparo podem ser enviados");
        }
        this.status = OrderStatus.OUT_FOR_DELIVERY;
    }
    
    public void markAsDelivered() {
        if (status != OrderStatus.OUT_FOR_DELIVERY) {
            throw new BusinessRuleException("MarkAsDelivered", "Apenas pedidos em entrega podem ser marcados como entregues");
        }
        this.status = OrderStatus.DELIVERED;
    }
    
    public void cancel() {
        if (status == OrderStatus.DELIVERED) {
            throw new BusinessRuleException("CancelOrder", "Pedidos já entregues não podem ser cancelados");
        }
        if (status == OrderStatus.CANCELLED) {
            throw new BusinessRuleException("CancelOrder", "Pedido já está cancelado");
        }
        this.status = OrderStatus.CANCELLED;
    }
    
    private void recalculateTotal() {
        this.totalValue = items.stream()
                .mapToDouble(OrderItem::getSubTotal)
                .sum();
    }
    
    private void validateCustomerId(UUID customerId) {
        if (customerId == null) {
            throw new InvalidEntityException("Order", "CustomerId não pode ser vazio");
        }
    }
    
    private void validateRestaurantId(UUID restaurantId) {
        if (restaurantId == null) {
            throw new InvalidEntityException("Order", "RestaurantId não pode ser vazio");
        }
    }
}